mgr/dashboard_v2: Import Ceph Model

sebastian-philipp opened the pull request:
This PR only imports the files to a state where the unit tests run through. 

It doesn't work yet.

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

rjfd reviewed (approved):
lgtm

votdev commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+        def check_base_type(val):
+            if not isinstance(val, self.base_type):
+                raise ValidationError(
+                    "invalid JSON type. Got {}, expected {}".format(type(parsed), self.base_type))
```
s/invalid/Invalid/

votdev commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+            parsed = json.loads(value)
+            return check_base_type(parsed)
+        except (ValueError, TypeError) as _:
+            raise ValidationError("invalid JSON")
```
s/invalid/Invalid/

votdev commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+class FloatField(JsonField):
+    def __init__(self, **kwargs):
+        super(FloatField, self).__init__(base_type=float, **kwargs)
+
```
Remove one of the 3 blank lines.

sebastian-philipp replied:
I took care of all linting errors today.

votdev commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+
+class NodbDoesNotExist(Exception):
+    pass
+
```
Add another blank line

votdev commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+        return LazyPropertyContributor(field_names, func)
+
+    return decorator
+
```
Add another blank line

sebastian-philipp replied:
I took care of all linting errors today.

votdev commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+    if class_dict is None:
+        class_dict = {}
+    return type(name, tuple(parents), class_dict)
+
```
Add another blank line

sebastian-philipp replied:
I took care of all linting errors today.

votdev commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+    def concrete_model(self):
+        return self.model
+
+    # make django filter happy.
```
I think this comment can be removed

LenzGr commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+
+    def iterator(self):
+        logger.warning(
+            '{}.iterator should only be access when running tests.'.format(self.__class__))
```
s/access/accessed/

votdev commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+    def add_field(self, field):
+        self.fields.append(field)
+        self.setup_pk(field)
+
```
Remove one of the 3 blank lines

sebastian-philipp replied:
I took care of all linting errors today.

votdev commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+
+        if 'objects' not in attrs:
+            bases[0].objects.copy_manager_to(new_class)
+
```
You will know what i mean :-)

sebastian-philipp replied:
I took care of all linting errors today.

LenzGr commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+    def add_field(self, field):
+        self.fields.append(field)
+        self.setup_pk(field)
+
```
One blank line too much?

sebastian-philipp replied:
I took care of all linting errors today.

votdev commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,897 @@
+
+        for k, w in kwargs.items():
+            self.__dict__[k] = w  # __dict__ is a dictproxy
+
```
^^^

sebastian-philipp commented:
@rjfd please review again.

p-se commented on src/pybind/mgr/dashboard_v2/models/cluster.py:
```diff
@@ -0,0 +1,160 @@
+
+class CephCluster(NodbModel, SendCommandApiMixin):
+    """Represents a Ceph cluster."""
+    # TODO: Lost functionallity compated to stand alone oA:
```
functionality compared

sebastian-philipp replied:
fixed

s0nea commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,892 @@
+
+def bulk_attribute_setter(field_names, catch_exceptions=None):
+    """
+    The idea @behind bulk_attribute_setter is to delay expensive calls to librados, until someone
```
Do you really want the '@' to be in front of 'behind'?

sebastian-philipp replied:
fixed

s0nea commented on src/pybind/mgr/dashboard_v2/models/nodb.py:
```diff
@@ -0,0 +1,892 @@
+    def save(self, force_insert=False, force_update=False, using=None,
+             update_fields=None):
+        """
+        This base implementation does nothing, except telling django that self is now successfully
```
Just a general question: I read 'django' serveral times now. Should we remove it on long term?

sebastian-philipp replied:
It makes sense to review all comments at a later stage, yes.

s0nea commented:
Maybe you already discussed it - in that case, sorry for raising it again. Otherwise what about the missing license headers?

rjfd commented on src/pybind/mgr/dashboard_v2/.pylintrc:
```diff
@@ -131,7 +131,11 @@ disable=print-statement,
         too-few-public-methods,
-        no-member
+        no-member,
+        bad-continuation,
```
this will make the code non-uniform across files, do you really need to disable this?

sebastian-philipp replied:
It forces a specific continuation style that was not enforced before. I've created https://tracker.openattic.org/browse/OP-3101

rjfd commented on src/pybind/mgr/dashboard_v2/.pylintrc:
```diff
@@ -330,7 +334,7 @@ indent-after-paren=4
 
 # Maximum number of characters on a single line.
-max-line-length=80
+max-line-length=100
```
In the Ceph project the max line length is 80, so we need to stick with it

rjfd commented on src/pybind/mgr/dashboard_v2/tox.ini:
```diff
@@ -31,4 +31,4 @@ commands =
 commands=
     pylint --rcfile=.pylintrc --jobs=5 . module.py tools.py ceph_module_mock.py controllers tests
-    pycodestyle --exclude=python2.7,.tox,venv,frontend .
+    pycodestyle --max-line-length=100 --exclude=python2.7,.tox,venv,frontend .
```
`--max-line-length=80`

sebastian-philipp committed:
Imported Nodb classes

* Some fixes were needed for Nodb.

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

sebastian-philipp committed:
Added Cluster Nodb model

Lots of fixes were needed for CephCluster, but
unfortunately, we don't have that many CephCluster tests.

It's not yet usable. The main goal was to fix the Python unit tests.

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

sebastian-philipp committed:
Added very basic and not yet usable REST controller for CephCluster

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

sebastian-philipp committed:
Skipped AuthTest.test_session_expire because it is too slow in my opiinion.

Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>
